File: 0152-maximum-product-subarray/0152-maximum-product-subarray.dart
class Solution {
  int maxProduct(List<int> nums) {
    int m = nums[0];
    int product = 1;
    for(int i=0;i<nums.length;i++){
        product *= nums[i];
        m = max(m, product);
        if(product == 0) product = 1;
    }
    product = 1;
    for(int i=nums.length-1;i>=0;i--){
        product *= nums[i];
        m = max(m, product);
        if(product == 0) product = 1;
    }
    return m;
  }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int maxProduct = nums[0];
        int minProduct = nums[0];
        int res = nums[0];
        for(int i=1;i<nums.size();i++){
            int x = nums[i];
            int newMax = max({x, maxProduct*x, minProduct*x});
            int newMin = min({x, maxProduct*x, minProduct*x});
            maxProduct = newMax;
            minProduct = newMin;
            res = max(res, maxProduct);
        }
        return res;
    }
};